### ISSUE TITLE
`broadcast(==` produces Vector{Real} and not Vector{Bool}

### ISSUE THREAD

**Issue by mauro3** [JuliaLang/julia#19218]
```julia
              _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.1116 (2016-10-25 02:18 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 9a2f0f1* (10 days old master)
|__/                   |  x86_64-pc-linux-gnu

julia> (==).(Real[1],[2])
1-element Array{Real,1}:
 false

julia> Real[1].==[2]
1-element BitArray{1}:
 false
```
Shouldn't both return the same type?  Preferably a `BitArray`.

**Comment by stevengj** [comment]
This is a duplicate of #17623; the `.==` functions are not yet handled by `broadcast`.  (That PR is currently stalled until `broadcast` functions with sparse arrays are improved; @Sacha0 is working on that.)

One difficulty with returning `BitArray` is that it doesn't work with fusion.  e.g. in #17623 `z .== y & y .> 0` will be fused into a single `broadcast` call, specialized `broadcast` methods for `==` etcetera are not called.

One option is to *always* return a `BitArray` from `broadcast` operations returning `Bool`.  However, it's not clear to me that this is a good idea.  Furthermore, given fusion, it seems that the advantage of bitarrays is greatly reduced: if you are producing multiple temporary arrays, a `BitArray` is much more attractive, combined with the fact that `.&` and `.|` operations for bitarrays can be faster than those operations on `Bool` arrays.

**Comment by mauro3** [comment]
Is this not also some promotion issue?  It works with `Number` (well it return a Vector{Bool} but that is fine too):
```
julia> (==).(Real[1//2, 1//2], [1,2])
2-element Array{Real,1}:
 false
 false

julia> (==).(Number[1//2, 1//2], [1,2])
2-element Array{Bool,1}:
 false
 false
```

**Comment by stevengj** [comment]
@mauro3, that is a separate issue.  cc @pabloferz

**Comment by pabloferz** [comment]
The `Vector{Real}` issue should be fixed by #18642.

**Comment by stevengj** [comment]
I think this can be closed now: as of #17623, `broadcast` always returns a `BitArray` for functions returning `Bool`.

**Comment by tkelman** [comment]
should add tests for that, if there aren't any yet
